File: frameworks/langchain/examples/rag-system.py
"""
LangChain RAG (Retrieval-Augmented Generation) System Example
Demonstrates how to build a document Q&A system with retrieval.
"""

import os
from langchain_openai import ChatOpenAI, OpenAIEmbeddings
from langchain_community.vectorstores import Chroma
from langchain_community.document_loaders import TextLoader, WebBaseLoader
from langchain_text_splitters import RecursiveCharacterTextSplitter
from langchain_core.prompts import ChatPromptTemplate
from langchain_core.output_parsers import StrOutputParser
from langchain_core.runnables import RunnablePassthrough

# Set up environment
# os.environ["OPENAI_API_KEY"] = "your-openai-api-key"

def format_docs(docs):
    """Format retrieved documents for the prompt."""
    return "\n\n".join([d.page_content for d in docs])

def create_sample_documents():
    """Create sample documents for demonstration."""
    
    documents = [
        {
            "content": """
            LangChain is a framework for developing applications powered by language models. 
            It enables applications that are context-aware and can reason. The framework 
            consists of several parts: LangChain Libraries, LangChain Templates, LangServe, 
            and LangSmith. It supports Python and JavaScript.
            """,
            "metadata": {"source": "langchain_intro", "topic": "framework"}
        },
        {
            "content": """
            CrewAI is designed for orchestrating role-playing, autonomous AI agents. 
            By fostering collaborative intelligence, CrewAI empowers agents to work 
            together seamlessly, tackling complex tasks. It's built in Python and 
            focuses on multi-agent coordination.
            """,
            "metadata": {"source": "crewai_intro", "topic": "multi-agent"}
        },
        {
            "content": """
            AutoGen is a framework that enables the development of LLM applications using 
            multiple agents that can converse with each other to solve tasks. It's developed 
            by Microsoft and supports various conversation patterns including two-agent, 
            group chat, and hierarchical structures.
            """,
            "metadata": {"source": "autogen_intro", "topic": "conversation"}
        },
        {
            "content": """
            Retrieval-Augmented Generation (RAG) is a technique that combines the power of 
            large language models with external knowledge sources. It retrieves relevant 
            information from a knowledge base and uses it to generate more accurate and 
            contextual responses.
            """,
            "metadata": {"source": "rag_explanation", "topic": "technique"}
        },
        {
            "content": """
            Vector databases are specialized databases designed to store and query 
            high-dimensional vectors efficiently. They are essential for RAG applications, 
            enabling semantic search over large document collections. Popular options 
            include Pinecone, Chroma, Weaviate, and Qdrant.
            """,
            "metadata": {"source": "vector_db_info", "topic": "database"}
        }
    ]
    
    return documents

def setup_vector_store():
    """Set up a vector store with sample documents."""
    
    print("📚 Setting up vector store...")
    
    # Create sample documents
    sample_docs = create_sample_documents()
    
    # Initialize text splitter
    text_splitter = RecursiveCharacterTextSplitter(
        chunk_size=500,
        chunk_overlap=50,
        length_function=len
    )
    
    # Create Document objects
    from langchain_core.documents import Document
    documents = []
    for doc in sample_docs:
        # Split the document if needed
        splits = text_splitter.split_text(doc["content"])
        for split in splits:
            documents.append(Document(
                page_content=split,
                metadata=doc["metadata"]
            ))
    
    # Initialize embeddings
    embeddings = OpenAIEmbeddings()
    
    # Create vector store
    vectorstore = Chroma.from_documents(
        documents=documents,
        embedding=embeddings,
        persist_directory="./chroma_db"
    )
    
    print(f"✅ Vector store created with {len(documents)} document chunks")
    return vectorstore

def create_rag_chain(vectorstore):
    """Create a RAG chain with retriever and generator."""
    
    # Initialize LLM
    llm = ChatOpenAI(
        model="gpt-4",
        temperature=0.3  # Lower temperature for more factual responses
    )
    
    # Create retriever
    retriever = vectorstore.as_retriever(
        search_type="similarity",
        search_kwargs={"k": 3}  # Retrieve top 3 most similar documents
    )
    
    # Create prompt template
    prompt = ChatPromptTemplate.from_messages([
        ("system", """You are an assistant for question-answering tasks. 
        Use the following pieces of retrieved context to answer the question. 
        If you don't know the answer based on the context, just say that you don't know. 
        Use three sentences maximum and keep the answer concise.
        
        Context: {context}"""),
        ("user", "{question}")
    ])
    
    # Create the RAG chain
    rag_chain = (
        {"context": retriever | format_docs, "question": RunnablePassthrough()}
        | prompt
        | llm
        | StrOutputParser()
    )
    
    return rag_chain, retriever

def example_basic_rag():
    """Basic RAG example with sample questions."""
    
    print("🔍 Basic RAG Example")
    print("=" * 40)
    
    # Set up vector store and RAG chain
    vectorstore = setup_vector_store()
    rag_chain, retriever = create_rag_chain(vectorstore)
    
    # Sample questions
    questions = [
        "What is LangChain and what does it do?",
        "How does CrewAI differ from other frameworks?",
        "What is RAG and why is it useful?",
        "What are vector databases used for?",
        "Which framework is best for multi-agent systems?"
    ]
    
    for question in questions:
        print(f"\n❓ Question: {question}")
        print("-" * 50)
        
        try:
            # Get answer from RAG chain
            answer = rag_chain.invoke(question)
            print(f"🤖 Answer: {answer}")
            
            # Show retrieved context
            docs = retriever.get_relevant_documents(question)
            print(f"\n📄 Retrieved {len(docs)} relevant documents:")
            for i, doc in enumerate(docs, 1):
                print(f"  {i}. Source: {doc.metadata.get('source', 'unknown')}")
                print(f"     Preview: {doc.page_content[:100]}...")
                
        except Exception as e:
            print(f"Error processing question: {e}")

def example_advanced_rag():
    """Advanced RAG example with custom retrieval and generation."""
    
    print("\n🎯 Advanced RAG Example")
    print("=" * 40)
    
    vectorstore = setup_vector_store()
    
    # Create custom retriever with different parameters
    retriever = vectorstore.as_retriever(
        search_type="mmr",  # Maximum Marginal Relevance
        search_kwargs={
            "k": 4,
            "fetch_k": 8,
            "lambda_mult": 0.7
        }
    )
    
    # Advanced LLM with specific instructions
    llm = ChatOpenAI(
        model="gpt-4",
        temperature=0.1,
        max_tokens=300
    )
    
    # Detailed prompt template
    prompt = ChatPromptTemplate.from_messages([
        ("system", """You are an expert AI framework consultant. Use the provided context to answer questions about AI agent frameworks.
        
        Guidelines:
        - Provide specific, actionable information
        - Compare frameworks when relevant
        - Include pros and cons when appropriate
        - Cite the source information when possible
        - If information is insufficient, clearly state what's missing
        
        Context:
        {context}
        
        Previous conversation:
        {chat_history}"""),
        ("user", "{question}")
    ])
    
    # Advanced chain with chat history
    rag_chain = (
        {
            "context": retriever | format_docs,
            "question": RunnablePassthrough(),
            "chat_history": lambda _: ""  # Simplified for this example
        }
        | prompt
        | llm
        | StrOutputParser()
    )
    
    # Complex questions
    complex_questions = [
        "I'm building a customer service bot that needs to handle complex multi-step conversations. Which framework should I choose and why?",
        "What are the trade-offs between using LangChain vs CrewAI for a document analysis system?",
        "How do I decide whether to use RAG or fine-tuning for my knowledge-intensive application?"
    ]
    
    for question in complex_questions:
        print(f"\n🤔 Complex Question: {question}")
        print("-" * 60)
        
        try:
            answer = rag_chain.invoke(question)
            print(f"💡 Expert Answer: {answer}")
            
        except Exception as e:
            print(f"Error processing complex question: {e}")

def example_web_rag():
    """Example: RAG with web content."""
    
    print("\n🌐 Web Content RAG Example")
    print("=" * 40)
    
    try:
        # Load web content
        urls = [
            "https://python.langchain.com/docs/introduction/",
            # Add more URLs as needed
        ]
        
        print("🔄 Loading web content...")
        
        # Note: This is a simplified example
        # In practice, you'd want to handle errors and timeouts
        documents = []
        
        # For demonstration, we'll use our sample documents
        # In a real scenario, you'd use WebBaseLoader
        print("📝 Using sample documents for demonstration...")
        vectorstore = setup_vector_store()
        rag_chain, _ = create_rag_chain(vectorstore)
        
        question = "What are the main components of LangChain?"
        answer = rag_chain.invoke(question)
        
        print(f"❓ Question: {question}")
        print(f"🌐 Answer from web content: {answer}")
        
    except Exception as e:
        print(f"Web RAG error: {e}")

def cleanup():
    """Clean up resources."""
    
    print("\n🧹 Cleaning up...")
    
    # Remove the Chroma database directory
    import shutil
    try:
        shutil.rmtree("./chroma_db")
        print("✅ Cleaned up vector database")
    except Exception as e:
        print(f"Cleanup warning: {e}")

def main():
    """Main function to run RAG examples."""
    
    print("🔍 LangChain RAG System Examples")
    print("=" * 45)
    
    # Check if API key is set
    if not os.getenv("OPENAI_API_KEY"):
        print("⚠️ Please set your OPENAI_API_KEY environment variable")
        print("Example: export OPENAI_API_KEY='your-api-key-here'")
        return
    
    try:
        # Run RAG examples
        example_basic_rag()
        
        input("\nPress Enter to continue to advanced RAG...")
        example_advanced_rag()
        
        input("\nPress Enter to continue to web RAG example...")
        example_web_rag()
        
        print("\n✅ All RAG examples completed!")
        
    except Exception as e:
        print(f"An error occurred: {e}")
        print("Make sure your API key is valid and you have internet connectivity.")
    
    finally:
        # Clean up
        cleanup()

if __name__ == "__main__":
    main()

File: frameworks/flowise/examples/rag-system.js
#!/usr/bin/env node
/**
 * RAG System Example - Flowise
 * =============================
 * 
 * This example demonstrates how to create a Retrieval-Augmented Generation (RAG) system
 * using Flowise. The system loads documents, creates embeddings, stores them in a vector
 * database, and answers questions based on the retrieved context.
 * 
 * Prerequisites:
 * - Node.js 18+ installed
 * - Flowise server running: npx flowise start
 * - OpenAI API key set in environment: OPENAI_API_KEY
 * - Sample documents in ./documents/ directory
 * 
 * Usage:
 *   node rag-system.js
 */

const fs = require('fs').promises;
const path = require('path');
const readline = require('readline');
const fetch = require('node-fetch');

class FlowiseRAGSystem {
    constructor(baseUrl = 'http://localhost:3000') {
        this.baseUrl = baseUrl;
        this.chatflowId = null;
        this.documentsPath = './documents';
        
        // Verify API key is set
        if (!process.env.OPENAI_API_KEY) {
            throw new Error('Please set OPENAI_API_KEY environment variable');
        }
    }
    
    /**
     * Create a RAG system flow configuration
     */
    createRAGFlowConfig() {
        return {
            name: "RAG Document Q&A System",
            flowData: JSON.stringify({
                nodes: [
                    // Document Loader Node
                    {
                        id: "folderFiles_0",
                        position: { x: 100, y: 100 },
                        type: "customNode",
                        data: {
                            id: "folderFiles_0",
                            label: "Folder Files",
                            name: "folderFiles",
                            type: "FolderFiles",
                            baseClasses: ["Document"],
                            inputs: {
                                folderPath: this.documentsPath,
                                recursive: true,
                                ignoredFiles: ".DS_Store"
                            }
                        }
                    },
                    
                    // Text Splitter Node
                    {
                        id: "recursiveCharacterTextSplitter_0",
                        position: { x: 400, y: 100 },
                        type: "customNode",
                        data: {
                            id: "recursiveCharacterTextSplitter_0",
                            label: "Recursive Character Text Splitter",
                            name: "recursiveCharacterTextSplitter",
                            type: "RecursiveCharacterTextSplitter",
                            baseClasses: ["RecursiveCharacterTextSplitter", "TextSplitter"],
                            inputs: {
                                chunkSize: 1000,
                                chunkOverlap: 200,
                                separators: "\\n\\n,\\n, ,."
                            }
                        }
                    },
                    
                    // OpenAI Embeddings Node
                    {
                        id: "openAIEmbeddings_0",
                        position: { x: 700, y: 100 },
                        type: "customNode",
                        data: {
                            id: "openAIEmbeddings_0",
                            label: "OpenAI Embeddings",
                            name: "openAIEmbeddings",
                            type: "OpenAIEmbeddings",
                            baseClasses: ["OpenAIEmbeddings", "Embeddings"],
                            inputs: {
                                modelName: "text-embedding-ada-002",
                                stripNewLines: true,
                                batchSize: 512,
                                timeout: 60000,
                                openAIApiKey: process.env.OPENAI_API_KEY
                            }
                        }
                    },
                    
                    // In-Memory Vector Store
                    {
                        id: "memoryVectorStore_0",
                        position: { x: 1000, y: 100 },
                        type: "customNode",
                        data: {
                            id: "memoryVectorStore_0",
                            label: "In-Memory Vector Store",
                            name: "memoryVectorStore",
                            type: "MemoryVectorStore",
                            baseClasses: ["MemoryVectorStore", "VectorStore"],
                            inputs: {}
                        }
                    },
                    
                    // Conversational Retrieval QA Chain
                    {
                        id: "conversationalRetrievalQAChain_0",
                        position: { x: 700, y: 400 },
                        type: "customNode",
                        data: {
                            id: "conversationalRetrievalQAChain_0",
                            label: "Conversational Retrieval QA Chain",
                            name: "conversationalRetrievalQAChain",
                            type: "ConversationalRetrievalQAChain",
                            baseClasses: ["ConversationalRetrievalQAChain", "BaseChain"],
                            inputs: {
                                returnSourceDocuments: true,
                                rephrasedQuestion: true,
                                responseTopK: 5,
                                systemMessagePrompt: `You are a helpful AI assistant. Use the following pieces of context to answer the question at the end.
If you don't know the answer, just say that you don't know, don't try to make up an answer.
Always provide the source of your information when possible.

{context}

Question: {question}
Helpful Answer:`
                            }
                        }
                    },
                    
                    // ChatOpenAI Model
                    {
                        id: "chatOpenAI_0",
                        position: { x: 400, y: 400 },
                        type: "customNode",
                        data: {
                            id: "chatOpenAI_0",
                            label: "ChatOpenAI",
                            name: "chatOpenAI",
                            type: "ChatOpenAI",
                            baseClasses: ["ChatOpenAI", "BaseChatModel"],
                            inputs: {
                                modelName: "gpt-4",
                                temperature: 0.1,
                                maxTokens: 1000,
                                topP: 1,
                                frequencyPenalty: 0,
                                presencePenalty: 0,
                                timeout: 60000,
                                openAIApiKey: process.env.OPENAI_API_KEY
                            }
                        }
                    }
                ],
                edges: [
                    // Document processing pipeline
                    {
                        id: "folderFiles_0-recursiveCharacterTextSplitter_0",
                        source: "folderFiles_0",
                        target: "recursiveCharacterTextSplitter_0",
                        sourceHandle: "folderFiles_0-output-folderFiles-Document",
                        targetHandle: "recursiveCharacterTextSplitter_0-input-documents-Document"
                    },
                    {
                        id: "recursiveCharacterTextSplitter_0-memoryVectorStore_0",
                        source: "recursiveCharacterTextSplitter_0",
                        target: "memoryVectorStore_0",
                        sourceHandle: "recursiveCharacterTextSplitter_0-output-recursiveCharacterTextSplitter-Document",
                        targetHandle: "memoryVectorStore_0-input-documents-Document"
                    },
                    {
                        id: "openAIEmbeddings_0-memoryVectorStore_0",
                        source: "openAIEmbeddings_0",
                        target: "memoryVectorStore_0",
                        sourceHandle: "openAIEmbeddings_0-output-openAIEmbeddings-OpenAIEmbeddings|Embeddings",
                        targetHandle: "memoryVectorStore_0-input-embeddings-Embeddings"
                    },
                    
                    // QA Chain connections
                    {
                        id: "memoryVectorStore_0-conversationalRetrievalQAChain_0",
                        source: "memoryVectorStore_0",
                        target: "conversationalRetrievalQAChain_0",
                        sourceHandle: "memoryVectorStore_0-output-retriever-VectorStoreRetriever",
                        targetHandle: "conversationalRetrievalQAChain_0-input-vectorStoreRetriever-VectorStoreRetriever"
                    },
                    {
                        id: "chatOpenAI_0-conversationalRetrievalQAChain_0",
                        source: "chatOpenAI_0",
                        target: "conversationalRetrievalQAChain_0",
                        sourceHandle: "chatOpenAI_0-output-chatOpenAI-ChatOpenAI|BaseChatModel",
                        targetHandle: "conversationalRetrievalQAChain_0-input-model-BaseChatModel"
                    }
                ]
            })
        };
    }
    
    /**
     * Create sample documents for testing
     */
    async createSampleDocuments() {
        try {
            await fs.mkdir(this.documentsPath, { recursive: true });
            
            const sampleDocs = {
                'ai_basics.txt': `
Artificial Intelligence (AI) is a branch of computer science that aims to create
intelligent machines that work and react like humans. AI systems can perform tasks
that typically require human intelligence, such as visual perception, speech
recognition, decision-making, and language translation.

There are three main types of AI:
1. Narrow AI (ANI) - AI that is designed to perform a narrow task
2. General AI (AGI) - AI that has generalized human cognitive abilities  
3. Super AI (ASI) - AI that surpasses human intelligence in all areas

Machine Learning is a subset of AI that provides systems the ability to
automatically learn and improve from experience without being explicitly programmed.

Deep Learning is a subset of machine learning that uses neural networks with
multiple layers to model and understand complex patterns in data.
                `,
                
                'flowise_guide.txt': `
Flowise is an open-source low-code tool for developers to build customized LLM
orchestration flows and AI agents. It provides a visual interface to create
complex AI workflows without extensive coding.

Key Features:
- Drag & drop interface for building AI flows
- Support for multiple LLM providers (OpenAI, Anthropic, etc.)
- Vector database integration for RAG applications
- Memory management for conversational agents
- Custom function calling capabilities
- API generation for created flows

Flowise is built on top of LangChain and provides a visual abstraction layer
that makes it easier to create production-ready AI applications.

Use Cases:
- Document Q&A systems
- Conversational AI chatbots
- Content generation workflows
- Data analysis and insights
- Customer support automation
                `,
                
                'rag_concepts.txt': `
Retrieval-Augmented Generation (RAG) is a technique that combines information
retrieval with text generation to create more accurate and contextual AI responses.

The RAG process involves several steps:
1. Document Ingestion - Loading and processing documents
2. Chunking - Breaking documents into manageable pieces
3. Embedding - Converting text chunks into vector representations
4. Storage - Storing embeddings in a vector database
5. Retrieval - Finding relevant chunks for a given query
6. Generation - Using retrieved context to generate responses

Benefits of RAG:
- Reduces hallucinations in AI responses
- Provides up-to-date information beyond training data
- Allows for source attribution and transparency
- Enables domain-specific knowledge integration
- Maintains context relevance for user queries

Vector databases commonly used with RAG:
- Pinecone (cloud-based)
- Chroma (open-source)
- Weaviate (open-source)
- Qdrant (open-source)
- FAISS (Facebook AI)
                `
            };
            
            for (const [filename, content] of Object.entries(sampleDocs)) {
                const filePath = path.join(this.documentsPath, filename);
                await fs.writeFile(filePath, content.trim());
            }
            
            console.log(`✅ Created ${Object.keys(sampleDocs).length} sample documents in ${this.documentsPath}`);
            
        } catch (error) {
            console.error('❌ Error creating sample documents:', error.message);
            throw error;
        }
    }
    
    /**
     * Create a new RAG chatflow in Flowise
     */
    async createRAGChatflow() {
        try {
            const config = this.createRAGFlowConfig();
            
            const response = await fetch(`${this.baseUrl}/api/v1/chatflows`, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify(config)
            });
            
            if (!response.ok) {
                const errorText = await response.text();
                throw new Error(`Failed to create RAG chatflow: ${response.statusText} - ${errorText}`);
            }
            
            const result = await response.json();
            this.chatflowId = result.id;
            
            console.log(`✅ RAG chatflow created with ID: ${this.chatflowId}`);
            
            // Wait a moment for the flow to initialize
            console.log('⏳ Initializing document processing...');
            await this.sleep(3000);
            
            return this.chatflowId;
            
        } catch (error) {
            console.error('❌ Error creating RAG chatflow:', error.message);
            throw error;
        }
    }
    
    /**
     * Query the RAG system with a question
     */
    async queryRAG(question, chatHistory = []) {
        try {
            if (!this.chatflowId) {
                throw new Error('No RAG chatflow created. Call createRAGChatflow() first.');
            }
            
            const response = await fetch(`${this.baseUrl}/api/v1/prediction/${this.chatflowId}`, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify({
                    question: question,
                    history: chatHistory,
                    overrideConfig: {}
                })
            });
            
            if (!response.ok) {
                const errorText = await response.text();
                throw new Error(`RAG query failed: ${response.statusText} - ${errorText}`);
            }
            
            const result = await response.json();
            
            return {
                answer: result.text || result.answer || 'No response received',
                sourceDocuments: result.sourceDocuments || [],
                chatHistory: result.chatHistory || chatHistory
            };
            
        } catch (error) {
            console.error('❌ Error querying RAG system:', error.message);
            return {
                answer: `Error: ${error.message}`,
                sourceDocuments: [],
                chatHistory: chatHistory
            };
        }
    }
    
    /**
     * Check document indexing status
     */
    async checkIndexingStatus() {
        try {
            // Try a simple query to see if documents are indexed
            const testResult = await this.queryRAG("What is AI?");
            
            if (testResult.sourceDocuments && testResult.sourceDocuments.length > 0) {
                console.log(`✅ Documents indexed successfully! Found ${testResult.sourceDocuments.length} relevant sources.`);
                return true;
            } else {
                console.log('⏳ Documents may still be indexing...');
                return false;
            }
            
        } catch (error) {
            console.log('⏳ Indexing in progress or not ready yet...');
            return false;
        }
    }
    
    /**
     * Display source information
     */
    displaySources(sourceDocuments) {
        if (!sourceDocuments || sourceDocuments.length === 0) {
            console.log('📚 No source documents found');
            return;
        }
        
        console.log(`\n📚 Sources (${sourceDocuments.length}):`);
        sourceDocuments.forEach((doc, index) => {
            const source = doc.metadata?.source || 'Unknown source';
            const content = doc.pageContent?.substring(0, 100) || 'No content';
            console.log(`  ${index + 1}. ${source}: "${content}..."`);
        });
    }
    
    /**
     * Check if Flowise server is running
     */
    async checkServerStatus() {
        try {
            const response = await fetch(`${this.baseUrl}/api/v1/ping`);
            return response.ok;
        } catch (error) {
            return false;
        }
    }
    
    /**
     * Clean up - delete the chatflow
     */
    async cleanup() {
        if (this.chatflowId) {
            try {
                await fetch(`${this.baseUrl}/api/v1/chatflows/${this.chatflowId}`, {
                    method: 'DELETE'
                });
                console.log('🧹 RAG chatflow cleaned up');
            } catch (error) {
                console.error('Warning: Failed to cleanup chatflow:', error.message);
            }
        }
    }
    
    /**
     * Sleep helper function
     */
    sleep(ms) {
        return new Promise(resolve => setTimeout(resolve, ms));
    }
}

/**
 * Interactive RAG system demo
 */
async function runInteractiveRAG() {
    const rag = new FlowiseRAGSystem();
    let chatHistory = [];
    
    // Setup readline interface
    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout,
        prompt: '\n❓ Question: '
    });
    
    console.log('\n🤖 RAG System - Ask questions about the documents!');
    console.log('Commands: "quit" to exit, "sources" to toggle source display');
    console.log('-'.repeat(60));
    
    let showSources = true;
    
    rl.prompt();
    
    rl.on('line', async (input) => {
        const question = input.trim();
        
        if (question.toLowerCase() === 'quit') {
            console.log('\n👋 Goodbye!');
            await rag.cleanup();
            rl.close();
            return;
        }
        
        if (question.toLowerCase() === 'sources') {
            showSources = !showSources;
            console.log(`📚 Source display: ${showSources ? 'ON' : 'OFF'}`);
            rl.prompt();
            return;
        }
        
        if (question) {
            console.log('🔍 Searching documents and generating answer...');
            
            const result = await rag.queryRAG(question, chatHistory);
            console.log(`\n🤖 Answer: ${result.answer}`);
            
            if (showSources) {
                rag.displaySources(result.sourceDocuments);
            }
            
            // Update chat history
            chatHistory = result.chatHistory;
        }
        
        rl.prompt();
    });
    
    rl.on('SIGINT', async () => {
        console.log('\n\n👋 Goodbye!');
        await rag.cleanup();
        process.exit(0);
    });
    
    return rag;
}

/**
 * Demonstration of various RAG queries
 */
async function demonstrateRAGQueries(rag) {
    console.log('\n🔄 Running demonstration queries...');
    console.log('='.repeat(50));
    
    const testQueries = [
        "What is Artificial Intelligence?",
        "What are the key features of Flowise?", 
        "How does RAG work?",
        "What are the benefits of using vector databases?",
        "What types of AI are there?"
    ];
    
    let chatHistory = [];
    
    for (const query of testQueries) {
        console.log(`\n❓ Question: ${query}`);
        console.log('🔍 Processing...');
        
        const result = await rag.queryRAG(query, chatHistory);
        
        console.log(`🤖 Answer: ${result.answer}`);
        
        if (result.sourceDocuments && result.sourceDocuments.length > 0) {
            console.log(`📚 Sources: ${result.sourceDocuments.length} documents found`);
        }
        
        chatHistory = result.chatHistory;
        
        // Small delay between queries
        await rag.sleep(1000);
    }
}

/**
 * Main function
 */
async function main() {
    console.log('🤖 Flowise RAG System Demo');
    console.log('='.repeat(50));
    
    const rag = new FlowiseRAGSystem();
    
    try {
        // Check if server is running
        console.log('🔍 Checking Flowise server...');
        const serverRunning = await rag.checkServerStatus();
        
        if (!serverRunning) {
            console.error('❌ Flowise server is not running!');
            console.log('💡 Start it with: npx flowise start');
            process.exit(1);
        }
        
        console.log('✅ Flowise server is running');
        
        // Create sample documents
        console.log('\n📚 Setting up sample documents...');
        await rag.createSampleDocuments();
        
        // Create RAG chatflow
        console.log('\n🏗️  Creating RAG chatflow...');
        await rag.createRAGChatflow();
        
        // Wait for document indexing
        console.log('\n⏳ Waiting for document indexing...');
        let indexed = false;
        let attempts = 0;
        const maxAttempts = 10;
        
        while (!indexed && attempts < maxAttempts) {
            await rag.sleep(2000);
            indexed = await rag.checkIndexingStatus();
            attempts++;
        }
        
        if (!indexed) {
            console.log('⚠️  Documents may still be indexing. Proceeding anyway...');
        }
        
        // Ask user what they want to do
        console.log('\n🎯 Choose an option:');
        console.log('1. Run interactive RAG session');
        console.log('2. Run demonstration queries');
        console.log('3. Both');
        
        const rl = readline.createInterface({
            input: process.stdin,
            output: process.stdout
        });
        
        rl.question('\nEnter choice (1-3): ', async (choice) => {
            rl.close();
            
            try {
                switch (choice.trim()) {
                    case '1':
                        await runInteractiveRAG();
                        break;
                    case '2':
                        await demonstrateRAGQueries(rag);
                        await rag.cleanup();
                        break;
                    case '3':
                        await demonstrateRAGQueries(rag);
                        await runInteractiveRAG();
                        break;
                    default:
                        console.log('Invalid choice. Running interactive session...');
                        await runInteractiveRAG();
                }
            } catch (error) {
                console.error('❌ Demo failed:', error.message);
                await rag.cleanup();
            }
        });
        
    } catch (error) {
        console.error('❌ Failed to start RAG system:', error.message);
        await rag.cleanup();
        process.exit(1);
    }
}

// Handle unhandled promise rejections
process.on('unhandledRejection', (reason, promise) => {
    console.error('Unhandled Rejection at:', promise, 'reason:', reason);
    process.exit(1);
});

// Start the application
if (require.main === module) {
    main().catch(console.error);
}

module.exports = FlowiseRAGSystem;